### PR TITLE
HOLY FUCKING SHIT, I'VE DONE IT! Map rotation!

### PR DESCRIPTION
**Pull request by MrsTonedOne**
This requires the use of the tgstation-server toolset

The current map is listed in the status and lobby stat panels (regardless of if map rotation is enabled)

Once the shuttle hits the point of no return, the ticker will roll for a chance to do a map rotation.

Rotate chance is calculated as ((round time in minutes)*0.75)%. So a 40minute round has a 30% chance of changing maps. a 60 minute round has a 45% chance.

The percentage part of that equation is configurable 

This is a **random** map rotation that is weighted based on player preferences. Server operators can also rig the weighting to favor certain maps.

Maps can have a min and max population.

A single configured default map will absorb the votes of people who haven't set a preferred map.

Renamed tgstation2 to Box Station

Adds two +SERVER admin verbs:
Force map rotation
Change Map
They do exactly what they say they do.

Adds a next map status tab entry.

### CODE REVIEW

**Comment by GunHog** [comment]
Lovely! :+1:

**Comment by Anonus** [comment]
VIVA MAP ROTATION

**Comment by Aloraydrel** [comment]
aaaaaaaaaah

**Comment by Alek2ander** [comment]
I hope this won't cause clients to redownload resources every map change.

**Comment by Wyzack** [comment]
Based MSO holy shit amazing 10/10

**Comment by MrStonedOne** [comment]
>Does this require the map to be merged for it to be selectable? It appears to, at least.

Yes, sadly. but no reason we can't merge some of these test maps in now.

**Comment by Cheridan** [comment]
:rotating_light: :rotating_light:

File: code/controllers/configuration.dm [hunk 1]
```diff
@@ -529,6 +535,58 @@
 	if(fps <= 0)
 		fps = initial(fps)
 
+
+/datum/configuration/proc/loadmaplist(filename)
+	var/list/Lines = file2list(filename)
```
Comment (Jordie0608) [inline code/controllers/configuration.dm]: You should use bygex for this. :shipit:

Comment (MrStonedOne) [inline code/controllers/configuration.dm]: one, too much work
two, too much work
three, with the amount of conditions i wanted to have, regex wouldn't have worked.

**Comment by hornygranny** [comment]
sign me the FUCK up  good shit go౦ԁ sHit thats ✔ some goodshit rightth  ere right✔there ✔✔if i do ƽaү so my selｆ 💯 i say so 💯 thats what im talking about right there right there (chorus: ʳᶦᵍʰᵗ ᵗʰᵉʳᵉ) mMMMMᎷМ💯  НO0ОଠＯOOＯOОଠଠOoooᵒᵒᵒᵒᵒᵒᵒᵒᵒ   💯     Good shit

**Comment by Supermichael777** [comment]
Mr Stoned one for pres-wait hes already our dictator.

**Comment by Dorsisdwarf** [comment]
@MrStonedOne 
>I hope this won't cause clients to redownload resources every map change.

**Comment by Metacide** [comment]
It's actually happening? My god. Thank you!

**Comment by MrStonedOne** [comment]
>I hope this won't cause clients to redownload resources every map change.
in theory, nothing should. Byond is suppose to be able to just download what's changed. Even if there is a difference of files or what not, it should be able to just grab the new or changed files.

In practice, any resource change that happens while another dream seeker window is open has a chance of doing that.

I've narrowed it down to either running ss13 that requires a partial resource download while downloading resources in another ss13 window (say, opening basil and sybil really quickly after an update)(100% chance). Or running ss13 while you have an old idle or disconnected ss13 window open where it had to download resources in the beginning (30% chance).

**Comment by Cookiesndip** [comment]
Question, if it rolls for a map rotation, can you 'rotate' to the current map? If not, that's shit.

**Comment by MrStonedOne** [comment]
Yes, the script has no care of what the current map is

**Comment by joonoro** [comment]
Nice! :+1:

**Comment by AlexanderUlanH** [comment]
As a powergaming box fanboy who has to know where all the maintenance loot spawners are on a map to play it this PR troubles me deeply.
:+1:

**Comment by daman997** [comment]
y e s :+1:

**Comment by Megatrononline** [comment]
What a horrible idea. All non-box maps give me claustrophobia. Not to mention that you need to get used to a map to play properly so a random map that you are not familiar with means it's just a ruined round for you. I suppose this is good for oldfags, but i play only for two years and what is this.

**Comment by Anonus** [comment]
If you're complaining a map rotation feature rotates the map, shoo.

**Comment by Megatrononline** [comment]
I mean if it will actually be used

**Comment by daman997** [comment]
>non-box maps give me claustrophodia
box is cramped as fuck compared to meta wat

**Comment by Incoming5643** [comment]
A good idea, but I still think we should just have a server that runs box 24/7 to placate those that enjoy the strong familiarity of it.

**Comment by Megatrononline** [comment]
Maybe something to do with a more horizontal layout as oppose to meta's more vertical layout. Or maybe something to do with that don't know it well enough to properly navigate. On box i feel freedom.

**Comment by Wyzack** [comment]
>Muh powergayman is slightly lessened on maps i do not know inside and out

welcome to the future nerds, we will drag you kicking and screaming if we have to

File: code/controllers/configuration.dm (hunk 1 above)
Comment (optimumtact) [inline code/controllers/configuration.dm]: :-1: don't be lazy

**Comment by Dorsisdwarf** [comment]
:+1: 
People should get out of the metahabits a server develops from only one map, but few people want to play other stations all the time.

**Comment by ike709** [comment]
:+1: maybe Basil will actually get populated if people can switch it to that ugly box station from time to time.

**Comment by Dorsisdwarf** [comment]
I think this PR deprecates basil actually

**Comment by optimumtact** [comment]
basil is dead because we no longer have the community size to support it, along with I assume improvements that made playing on a higher pop server less horrible. So the only way to fix it is to attract enough new players to make it worthwhile, perhaps the next summer rush.

**Comment by Shadowlight213** [comment]
Are admins able to force set a new map?

**Comment by pudl** [comment]
:-1: There's no purpose.

**Comment by Iamgoofball** [comment]
@pudl yes there is and you know it

what is "every single other multiplayer game in existence"

**Comment by pudl** [comment]
SS13 isn't a normal multiplayer game. Unless you're looking at absolutely batshit tactical precision, SS13 is a game in which the map you're playing on affects the game more than any other single thing.

**Comment by Iamgoofball** [comment]
Every other game has a different play style and system depending on the map. 

Name a few multiplayer games where the map does NOT change how you play the game in the setting.

**Comment by pudl** [comment]
Garry's Mod, Minecraft, Euro Truck Simulator 2, Borderlands, every fighting game ever.

**Comment by Iamgoofball** [comment]
>garry's mod

gmod isn't a game, it's a platform
even if you are to argue the sandbox bit

your building space/what you build is heavily decided by your map you are on

>Minecraft

you serious right now? the world you spawn in completely determines what you do and what resources you have

>Euro Truck Simulator 2

The route you drive along and how long it takes you along with traffic congestion is decided by the map you are on. A map of the Red Faction mars base with cars and shit would have a different layout for you to work with and different short cuts than europe.

>Borderlands

This is an FPS. Level Design in FPS is one of the most important things. Bad level design can make or break an FPS game. Your cover options, enemies, etc. are all decided by the map. The loot spawns, one of the most important things in the game, is also decided by the map.

>Every fighting game ever

Alright, you got me there. That's a good example of a few. However, SS13 isn't a closed up fighting game where the map is purely a background.

File: code/controllers/subsystem/ticker.dm [hunk 2]
```diff
@@ -458,4 +461,79 @@ var/datum/subsystem/ticker/ticker
 		if(25 to INFINITY)  //No response from the next in line when a vacancy exists, remove he
 			next_in_line << "<span class='danger'>No response recieved. You have been removed from the line.</span>"
 			queued_players -= next_in_line
-			queue_delay = 0
\ No newline at end of file
+			queue_delay = 0
+
+/datum/subsystem/ticker/proc/check_maprotate(var/force = 0)
```
Comment (Shadowlight213) [inline code/controllers/subsystem/ticker.dm]: The force argument for this proc looks like it got left over from the debug stuff, as its not used anywhere.

**Comment by Mloc** [comment]
be careful with using shell()
it's known to lock up BYOND horribly

**Comment by MrStonedOne** [comment]
@mloc It seems to be fine now, shell() is used by the bot to notify the admin channel ALOT.

Every time the shuttle docks at the station
Every time the round ends.
Every time somebody on the watch list joins when no admins are on
Every time an admin help comes in when no admins are on
Every time a first time player joins when no admins are on

File: code/controllers/subsystem/ticker.dm (hunk 2 above)
Comment (MrStonedOne) [inline code/controllers/subsystem/ticker.dm]: Thank you fixing.

**Comment by MrStonedOne** [comment]
@Shadowlight213 not currently. But I'll look at adding a verb or some shit.

**Comment by Dorsisdwarf** [comment]
3 hours.

**Comment by Mloc** [comment]
it might just be linux actually- and since this calls bat files you're already breaking platform compatibility :p

**Comment by MrStonedOne** [comment]
you can run bat files in linux!

They would have to contain a shebang and bash code, but its doable =P

**Comment by Strathcona** [comment]
Damn this is awesome! Will this make testing maps like dreamstation or whatever the mapping section spits out easier?

**Comment by Razharas** [comment]
@MrStonedOne 
Tell when this is ready

**Comment by Dorsisdwarf** [comment]
Live test when

**Comment by MrStonedOne** [comment]
>Live test when
Coming up. I'm gonna make another PR with all 3 maps to be tested, and this change as well. Then I'll close the pr and do a live test of it.

**Comment by MrStonedOne** [comment]
#12111

**Comment by Metacide** [comment]
Incredible! It works very well, and seamlessly too by the looks of things. I never really thought anyone'd get round to doing this. A bit more motivation to mappers, too. Thanks very much, really! :+1:

**Comment by Strathcona** [comment]
I agree with Meta here, this is probably one of the best improvements to the core game in a looooong time.

File: code/world.dm [hunk 3]
```diff
@@ -294,3 +297,77 @@ var/failed_db_connections = 0
...
+		mapvotes[map] = mapvotes[map]*VM.voteweight
+
+	var/pickedmap = pickweight(mapvotes)
+	if (!pickedmap)
+		return
+	var/datum/votablemap/VM = config.maplist[pickedmap]
+	message_admins("Randomly rotating map to [VM.name]([VM.friendlyname])")
+	. = changemap(VM)
+	if (. == 0)
+		world << "<span class='boldannounce'>Map rotation has choosen [VM.friendlyname] for next round!</span>"
```
Comment (freerealestate) [inline code/world.dm]: it's uh

'chosen'

File: code/modules/admin/verbs/maprotation.dm [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+				mapname += "0"
+			mapname += "-"
+			if (VM.maxusers > 0)
+				mapname += "[VM.maxusers]"
+			else
+				mapname += "inf"
+			mapname += "\]"
+
+		maprotatechoices[mapname] = VM
+	var/choosenmap = input("Choose a map to rotate to", "Change Map")  as null|anything in maprotatechoices
```
Comment (freerealestate) [inline code/modules/admin/verbs/maprotation.dm]: chosen

**Comment by Jordie0608** [comment]
An aside suggestion: add an indicator to Show Server Revision if map rotation is enabled or not.

**Comment by Razharas** [comment]
@MrStonedOne 
Is this ready?

**Comment by MrStonedOne** [comment]
No, It has some bugs still. I'll have them mostly fixed. I'm trying to figure out how to fix the last one in a way that doesn't suck.

**Comment by MrStonedOne** [comment]
This is as ready as it will ever be for merge.

**Comment by Giacom** [comment]
I'm not clear on how this works. I'm currently guessing it does this:

* Shuttle hits point of no return.
* Server starts compiling a new game with the new map.
* The current round ends.
* Server restarts and loads the new map.

If so, then is there a long downtime between rounds?

**Comment by GunHog** [comment]
So far it has been pretty seamless.

**Comment by Anonus** [comment]
IT IS DONE LOOK AT ALL THESE TWO MAPS IN ROTATION

**Comment by Jordie0608** [comment]
There are four.

**Comment by freerealestate** [comment]
where's asteroid and uterus tho

**Comment by Shadowlight213** [comment]
Asteroid station is completely broken and was deleted from the repo.

**Comment by Giacom** [comment]
https://github.com/tgstation/-tg-station/blob/master/config/maps.txt says 3 maps.

**Comment by RemieRichards** [comment]
@Giacom it's a .txt config file, they're almost ALWAYS different on the live servers to the one on the repo.

**Comment by freerealestate** [comment]
>Asteroid station is completely broken and was deleted from the repo.

RIP

What was broken about it, out of curiosity?

**Comment by Shadowlight213** [comment]
It ended up even failing to load in dreammaker, which meant it was impossible to continue work on.

**Pull request merged**